### PR TITLE
feat: use a generated intermediate address as the forwarder

### PR DESCRIPTION
**Pull request by javiersuweijie**
This PR replaces https://github.com/cosmos/ibc-apps/pull/69. Instead of enforce the usage of the module account as the intermediate forwarder, PMF will generate an address based on the sender's address.

This method has the following advantages:
1. Receiver is able to verify where the message came from by perform the intermediate address generation method
2. When a chain does not have PFM installed, the message will fallback to the receiving account which is assumed to be controlled by the sender.

### CODE REVIEW

File: middleware/packet-forward-middleware/router/keeper/keeper.go [hunk 1]
```diff
@@ -218,7 +219,7 @@ func (k *Keeper) ForwardTransferPacket(
 
 	// pay fees
 	if feeAmount.IsPositive() {
-		hostAccAddr, err := sdk.AccAddressFromBech32(receiver)
+		hostAccAddr, err := sdk.AccAddressFromBech32(feePayer)
```
Comment (agouin) [inline middleware/packet-forward-middleware/router/keeper/keeper.go:222]: the `intermediateSender` should be the fee payer here, as that is where the funds are received, and will be forwarded out of.

File: middleware/packet-forward-middleware/router/keeper/keeper.go [hunk 2]
```diff
@@ -391,12 +393,21 @@ func (k *Keeper) RetryTimeout(
 
 	token := sdk.NewCoin(denom, amount)
 
+	// Handle backward compatibility for inflight packets right after the middleware upgrade
```
Comment (agouin) [inline middleware/packet-forward-middleware/router/keeper/keeper.go:396]: this is not necessary when the fee payer is the `intermediateSender`

File: middleware/packet-forward-middleware/router/keeper/keeper.go [hunk 3]
```diff
@@ -202,7 +202,8 @@ func (k *Keeper) ForwardTransferPacket(
 	inFlightPacket *types.InFlightPacket,
 	srcPacket channeltypes.Packet,
 	srcPacketSender string,
-	receiver string,
+	feePayer string,
```
Comment (agouin) [inline middleware/packet-forward-middleware/router/keeper/keeper.go:205]: we shouldn't need to separate `feePayer` and `intermediateSender`

**Review by agouin — CHANGES_REQUESTED** [review]

File: middleware/packet-forward-middleware/router/types/params.go [hunk 4]
```diff
@@ -10,7 +10,8 @@ import (
 var (
 	DefaultFeePercentage = sdk.NewDec(0)
 	// KeyFeePercentage is store's key for FeePercentage Params
-	KeyFeePercentage = []byte("FeePercentage")
+	KeyFeePercentage       = []byte("FeePercentage")
+	IntermediateAddrPrefix = "packet-fowarder"
```
Comment (agouin) [inline middleware/packet-forward-middleware/router/types/params.go:14]:
```suggestion
	IntermediateAddrPrefix = "packet-forwarder"
```

File: middleware/packet-forward-middleware/router/keeper/keeper.go (hunk 3 above)
Comment (nicolaslara) [inline middleware/packet-forward-middleware/router/keeper/keeper.go:205]: Why is this called fee payer? AFAIK the fee is paid by the relayer everywhere except for the initial send

Comment (nicolaslara) [reply middleware/packet-forward-middleware/router/keeper/keeper.go:205]: oh, I see! this is in the case where there is a forwarding fee. Interesting 🤔 I wonder if any chain has the fees enabled.

**Comment by javiersuweijie** [comment]
Thanks @agouin. Somehow I thought the fee payer was going to pay in the native token existing in the account instead of using the amount that was transferred. This makes a lot more sense. Thanks for the review.

**Comment by javiersuweijie** [comment]
Since the changes here are the same as https://github.com/cosmos/ibc-apps/pull/71 closing this in favour of that.

**Pull request closed without merging**